### PR TITLE
Add support for FC WWPN candidates

### PR DESCRIPTION
**Pull request by borisko123**
Support for FC WWPN candidates

Depends on:
- [x] ManageIQ/manageiq-schema#596
- [x] ManageIQ/manageiq#21363

### CODE REVIEW

File: app/models/manageiq/providers/autosde/inventory/collector/storage_manager.rb [hunk 1]
```diff
@@ -73,4 +73,17 @@ def physical_storage_families
       }
     end
   end
+
+  def storage_hosts_wwpn_candidates
+    @storage_hosts_wwpn_candidates ||= @manager.autosde_client.StorageHostWWPNCandidatesApi
+                                               .storage_hosts_wwpn_candidates_get.map do |candidate|
+      {
+        :candidates => candidate.wwpn,
```
Comment (agrare) [inline app/models/manageiq/providers/autosde/inventory/collector/storage_manager.rb:81]: Minor formatting issue but can you line up the ` =>` with the hash rockets below?
E.g.
```ruby
{
  :candidates          => candidate.wwpn,
  :physical_storage_id => candidate.system_uuid,
  :ems_ref             => candidate_system_uuid
}
```

File: app/models/manageiq/providers/autosde/inventory/collector/storage_manager.rb [hunk 2]
```diff
@@ -73,4 +73,17 @@ def physical_storage_families
...
+        :physical_storage_id => candidate.system_uuid,
+        :ems_ref => candidate.system_uuid
+
```
Comment (agrare) [inline app/models/manageiq/providers/autosde/inventory/collector/storage_manager.rb:84]: Extra newline here

File: app/models/manageiq/providers/autosde/inventory/collector/storage_manager.rb [hunk 3]
```diff
@@ -73,4 +73,17 @@ def physical_storage_families
       }
     end
   end
+
+  def storage_hosts_wwpn_candidates
+    @storage_hosts_wwpn_candidates ||= @manager.autosde_client.StorageHostWWPNCandidatesApi
+                                               .storage_hosts_wwpn_candidates_get.map do |candidate|
+      {
+        :candidates => candidate.wwpn,
+        :physical_storage_id => candidate.system_uuid,
```
Comment (agrare) [inline app/models/manageiq/providers/autosde/inventory/collector/storage_manager.rb:82]: I think you want to do a `persister.physical_storages.lazy_find(candidate.system_uuid)` here.  Weird that it is in the collector since this is typically done in the parser.

File: app/models/manageiq/providers/autosde/inventory/parser/storage_manager.rb [hunk 4]
```diff
@@ -107,4 +109,10 @@ def cloud_volumes
       )
     end
   end
+
+  def storage_host_wwpn_candidates
+      collector.storage_hosts_wwpn_candidates.each do |candidate|
+        persister.storage_hosts_wwpn_candidates.build(**candidate)
```
Comment (agrare) [inline app/models/manageiq/providers/autosde/inventory/parser/storage_manager.rb:115]: I recommend having your collector just return the `candidate` like:

```ruby
def storage_hosts_wwpn_candidates
  @storage_hosts_wwpn_candidates ||= @manager.autosde_client.StorageHostWWPNCandidatesApi
                                             .storage_hosts_wwpn_candidates_get
end
```

then your parser would look like
```ruby
def storage_host_wwpn_candidates
  collector.storage_hosts_wwpn_candidates.each do |candidate|
    persister.storage_hosts_wwpn_candidates.build(
      :candidates       => candidate.wwpn,
      :ems_ref          => candidate.system_uuid,
      :physical_storage => persister.physical_storages.lazy_find(candidate.system_uuid)

    )
  end
end
```

File: app/models/manageiq/providers/autosde/inventory/persister/storage_manager.rb [hunk 5]
```diff
@@ -65,6 +65,11 @@ def initialize_inventory_collections
         :ems_id => ->(persister) { persister.manager.id }
       )
     end
+    # wwpn candidates
+    add_collection(storage, :storage_hosts_wwpn_candidates) do |builder|
+      builder.add_default_values(
+        :ems_id => ->(persister) { persister.manager.id },)
```
Comment (agrare) [inline app/models/manageiq/providers/autosde/inventory/persister/storage_manager.rb:71]: the trailing `)` should be on a newline and the trailing `,` isn't necessary

File: spec/models/manageiq/providers/autosde/inventory_spec.rb [hunk 6]
```diff
@@ -23,5 +23,7 @@
     expect(ems.cloud_volumes).to_not be_empty
 
     expect(ems.cloud_volumes.first).to be_instance_of(ManageIQ::Providers::Autosde::StorageManager::CloudVolume)
+    expect(ems.storage_hosts_wwpn_candidates).to_not be_empty
+
```
Comment (agrare) [inline spec/models/manageiq/providers/autosde/inventory_spec.rb:27]: Extra newline

File: spec/models/manageiq/providers/autosde/inventory_spec.rb [hunk 7]
```diff
@@ -23,5 +23,7 @@
     expect(ems.cloud_volumes).to_not be_empty
 
     expect(ems.cloud_volumes.first).to be_instance_of(ManageIQ::Providers::Autosde::StorageManager::CloudVolume)
+    expect(ems.storage_hosts_wwpn_candidates).to_not be_empty
```
Comment (agrare) [inline spec/models/manageiq/providers/autosde/inventory_spec.rb:26]: Ideally we should be checking the attributes for one of the records that gets created.  This helps ensure that e.g. the lazy_find association was built correctly

File: spec/models/manageiq/providers/autosde/wwpn_candidates_spec.rb [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    ems = FactoryBot.create(:autosde_storage_manager, :with_autosde_credentials, :hostname => Rails.application.secrets.autosde[:appliance_host])
+
+    VCR.use_cassette("get_storage_systems_wwpn_candidates", :record => :once) do
+      client = ManageIQ::Providers::Autosde::StorageManager::AutosdeClient.new(
+        :host     => Rails.application.secrets.autosde[:appliance_host],
+        :username => Rails.application.secrets.autosde[:site_manager_user],
+        :password => Rails.application.secrets.autosde[:site_manager_password],
+        :scheme => 'http'
+      )
+      result =  client.StorageHostWWPNCandidatesApi.storage_hosts_wwpn_candidates_get
```
Comment (agrare) [inline spec/models/manageiq/providers/autosde/wwpn_candidates_spec.rb:13]: Some extra padding after the `result =     `

File: spec/models/manageiq/providers/autosde/wwpn_candidates_spec.rb [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+      expect(result).to be_an_instance_of(Array)
```
Comment (agrare) [inline spec/models/manageiq/providers/autosde/wwpn_candidates_spec.rb:14]: Probably should add extra tests than just returns an array

**Comment by borisko123** [comment]
@miq-bot cross_repo_test ManageIQ/manageiq#21363 ManageIQ/manageiq-schema#596

**Comment by borisko123** [comment]
@miq-bot cross_repo_test ManageIQ/manageiq#21363

**Comment by borisko123** [comment]
@miq-bot cross_repo_test ManageIQ/manageiq#21363, ManageIQ/manageiq-schema#596

**Comment by borisko123** [comment]
@miq-bot cross_repo_test ManageIQ/manageiq#21363

File: app/models/manageiq/providers/autosde/inventory/persister/storage_manager.rb [hunk 10]
```diff
@@ -65,6 +65,12 @@ def initialize_inventory_collections
         :ems_id => ->(persister) { persister.manager.id }
       )
     end
+    # wwpn candidates
+    add_collection(storage, :wwpn_candidates) do |builder|
+      builder.add_default_values(
+        :ems_id => ->(persister) { persister.manager.id }
+      )
```
Comment (agrare) [inline app/models/manageiq/providers/autosde/inventory/persister/storage_manager.rb:72]: This should be handled by the `add_common_default_values` in the core persister definitions https://github.com/ManageIQ/manageiq/pull/21363/files#diff-f623f973e49e35b8e6526b8a41057bb3707c90ed14623c932e38054ec081fb3dR34-R36

It isn't harming anything so I'll merge with this here but can you put in a follow-up PR to remove this?

File: app/models/manageiq/providers/autosde/storage_manager/cloud_volume.rb [hunk 11]
```diff
@@ -1,8 +1,13 @@
 class ManageIQ::Providers::Autosde::StorageManager::CloudVolume < ::CloudVolume
   supports :create
-  supports :safe_delete
+  supports :delete do
+    unsupported_reason_add(:delete, _("the volume is not connected to an active provider")) unless ext_management_system
+  end
+  supports :safe_delete do
+    unsupported_reason_add(:safe_delete, _("the volume is not connected to an active provider")) unless ext_management_system
+  end
```
Comment (agrare) [inline app/models/manageiq/providers/autosde/storage_manager/cloud_volume.rb:8]: Is this related to the wwpn_candidates change?

Comment (agrare) [reply app/models/manageiq/providers/autosde/storage_manager/cloud_volume.rb:8]: Oh you brought in other commits from master, please rebase properly

**Review by agrare — DISMISSED** [review]
Need to remove all merge commits from this branch e.g. https://github.com/ManageIQ/manageiq-providers-autosde/pull/74/commits/c69185427216e2bb77e055a6b1107e8c0dc18cc0 pulled in @chessbyte's cloud_volume supports feature mixin work

File: spec/models/manageiq/providers/autosde/inventory_spec.rb [hunk 12]
```diff
@@ -23,5 +23,11 @@
     expect(ems.cloud_volumes).to_not be_empty
 
     expect(ems.cloud_volumes.first).to be_instance_of(ManageIQ::Providers::Autosde::StorageManager::CloudVolume)
+
+    # wwpn candidates
+    expect(ems.wwpn_candidates).to_not be_empty
```
Comment (agrare) [inline spec/models/manageiq/providers/autosde/inventory_spec.rb:28]: Better than `expect().not_to be_empty` would be to check `expect(ems.wwpn_candidates.count).to eq(whatever)`

File: spec/models/manageiq/providers/autosde/inventory_spec.rb [hunk 13]
```diff
@@ -23,5 +23,11 @@
...
+    first_candidate = ems.wwpn_candidates.first
```
Comment (agrare) [inline spec/models/manageiq/providers/autosde/inventory_spec.rb:29]: :-1: for using `.first` unless you also explicitly expect the count to be `1` above, otherwise you're going to hit sporadic test failures.

Prefer to use `ems.wwpn_candidates.find_by(:ems_ref => "something")`

File: spec/models/manageiq/providers/autosde/inventory_spec.rb [hunk 14]
```diff
@@ -23,5 +23,11 @@
...
+    expect(first_candidate.candidate).not_to be_empty
```
Comment (agrare) [inline spec/models/manageiq/providers/autosde/inventory_spec.rb:30]: Best to do:
```
expect(first_candidate).to have_attributes(
  :candidate        => "something",
  :ems_ref          => "something_else",
  :physical_storage => ems.physical_storages.find_by(:ems_ref => "whatever")
)
```

**Comment by borisko123** [comment]
> Need to remove all merge commits from this branch e.g. [c691854](https://github.com/ManageIQ/manageiq-providers-autosde/commit/c69185427216e2bb77e055a6b1107e8c0dc18cc0) pulled in @chessbyte's cloud_volume supports feature mixin work

done

File: spec/models/manageiq/providers/autosde/inventory_spec.rb [hunk 15]
```diff
@@ -5,23 +5,35 @@
...
+    expect(ems.cloud_volumes.first).to(be_instance_of(ManageIQ::Providers::Autosde::StorageManager::CloudVolume))
 
-    expect(ems.cloud_volumes.first).to be_instance_of(ManageIQ::Providers::Autosde::StorageManager::CloudVolume)
+    # wwpn candidates
+    expect(ems.wwpn_candidates.count).to(eq(2))
+    first_candidate = ems.wwpn_candidates.first
+    expect(first_candidate).to(have_attributes(
+                                 :candidate        => first_wwpn,
+                                 :ems_id           => ems.id,
+                                 :physical_storage => ems.physical_storages.find_by(:ems_id => ems.id)
```
Comment (agrare) [inline spec/models/manageiq/providers/autosde/inventory_spec.rb:35]: `ems.physical_storage.find_by(:ems_id => ems.id)` is redudant (ems.physical_storages already does that), think you meant to do `.find_by(:ems_ref => "ref-of-your-physical-storage")`

File: spec/models/manageiq/providers/autosde/inventory_spec.rb (hunk 13 above)
Comment (agrare) [reply spec/models/manageiq/providers/autosde/inventory_spec.rb:31]: Not sure why this was resolved, still an issue

File: spec/models/manageiq/providers/autosde/inventory_spec.rb (hunk 15 above)
Comment (borisko123) [reply spec/models/manageiq/providers/autosde/inventory_spec.rb:35]: For wwpn_candidates, we never going to create them, only read them
But persister need some uniq key, when it builds InventoryObject in InventoryCollection. I needed workaround.
I've recalled I was using ems_ref as such key, and gave it the candidate value (it is uniq wordlwide).
So in the test I need only to check if wwpn_candidate is created as expected, and not to use ems_ref for finding anything
Thus, knowing ems or physical_storage,
I can get wwpm_candidates like:
* ems.wwpn_candidates
* ems.physical_storages.first.wwpn_candidates


From app/models/manageiq/providers/autosde/inventory/parser/storage_manager.rb:

`def wwpn_candidates
    collector.wwpn_candidates.each do |candidate|
      persister.wwpn_candidates.build(
        :candidate        => candidate.wwpn,
        :ems_ref          => candidate.wwpn,
        :physical_storage => persister.physical_storages.lazy_find(candidate.system_uuid)
      )
    end
  end`

Comment (agrare) [reply spec/models/manageiq/providers/autosde/inventory_spec.rb:35]: Sorry not sure what you mean.  You're doing `ems.physical_storages.find_by(:ems_id => ems.id)` which doesn't make any sense.  If you have multiple physical storages and you need to find one (and find the same one consistently) then you need to `.find_by(:ems_ref => "ref-of-the physical-storage")`

Comment (agrare) [reply spec/models/manageiq/providers/autosde/inventory_spec.rb:35]: High level, this spec is just querying the database.  It has nothing to do with creating or reading wwpn_candidates, all of that has already happened in the refresher.

Now you just need to query the DB to find the same records consistently so that your specs don't fail randomly.  That's the problem with using `first_candidate = ems.wwpn_candidates.first`

File: spec/models/manageiq/providers/autosde/inventory_spec.rb [hunk 16]
```diff
@@ -5,23 +5,35 @@
...
+                               ))
+    expect(first_candidate.physical_storage).to(eq(ems.physical_storages.first))
```
Comment (agrare) [inline spec/models/manageiq/providers/autosde/inventory_spec.rb:37]:
```suggestion
    expect(ems.wwpn_candidates.count).to eq(2)
    first_candidate = ems.wwpn_candidates.find_by(:ems_ref => "2100000E1EE89D90")
    expect(first_candidate).to have_attributes(
      :candidate        => first_wwpn,
      :physical_storage => ems.physical_storages.find_by(:ems_ref => "a785c49a-2f3c-4e6d-8ace-879dfa1719a2")
    )
```

**Review by agrare — CHANGES_REQUESTED** [review]

File: spec/models/manageiq/providers/autosde/inventory_spec.rb [hunk 17]
```diff
@@ -5,23 +5,34 @@
                       :hostname => Rails.application.secrets.autosde[:appliance_host])
   end
 
+  let(:first_wwpn) { "2100000E1EE89D90" }
+  let(:ems_ref) { "a785c49a-2f3c-4e6d-8ace-879dfa1719a2" }
```
Comment (agrare) [inline spec/models/manageiq/providers/autosde/inventory_spec.rb:9]: Don't want to drag this out any longer, can be fixed in a follow-up but this should be `:physical_storage_ems_ref` or something else that indicates what it is for

**Review by agrare — APPROVED** [review]
